File: lib/mercuryapi-1.23.0.20/java/samples_nb/src/samples/SecureReadData.java
/**
 * Sample program that reads tags in the background and prints the
 * tags found with corresponding password.
 */
// Import the API
package samples;

import com.thingmagic.*;
import java.text.SimpleDateFormat;
import java.util.Calendar;

public class SecureReadData
{

    static SerialPrinter serialPrinter;
    static StringPrinter stringPrinter;
    static TransportListener currentListener;

    static void usage()
    {
        System.out.printf("Usage: demo reader-uri <command> [args]\n"
                + "  (URI: 'tmr:///COM1' or 'tmr://astra-2100d3/' "
                + "or 'tmr:///dev/ttyS0')\n\n"
                + "Available commands:\n");
        System.exit(1);
    }

    public static void setTrace(Reader r, String args[])
    {
        if (args[0].toLowerCase().equals("on"))
        {
            r.addTransportListener(Reader.simpleTransportListener);
            currentListener = Reader.simpleTransportListener;
        }
        else if (currentListener != null)
        {
            r.removeTransportListener(Reader.simpleTransportListener);
        }
    }

    static class SerialPrinter implements TransportListener
    {
        public void message(boolean tx, byte[] data, int timeout)
        {
            System.out.print(tx ? "Sending: " : "Received:");
            for (int i = 0; i < data.length; i++)
            {
                if (i > 0 && (i & 15) == 0)
                {
                    System.out.printf("\n         ");
                }
                System.out.printf(" %02x", data[i]);
            }
            System.out.printf("\n");
        }
    }

    static class StringPrinter implements TransportListener
    {
        public void message(boolean tx, byte[] data, int timeout)
        {
            System.out.println((tx ? "Sending:\n" : "Receiving:\n")
                    + new String(data));
        }
    }


    public static void main(String argv[])
    {
        // Program setup
        Reader reader = null;
        int nextarg = 0;
        boolean trace = false;

        if (argv.length < 1)
        {
            usage();
        }

        if (argv[nextarg].equals("-v"))
        {
            trace = true;
            nextarg++;
        }

        // Create Reader object, connecting to physical device
        try
        {
            reader = Reader.create(argv[nextarg]);
            if (trace)
            {
                setTrace(reader, new String[]{"on"});
            }
            reader.connect();
            if (Reader.Region.UNSPEC == (Reader.Region) reader.paramGet("/reader/region/id"))
            {
                Reader.Region[] supportedRegions = (Reader.Region[]) reader.paramGet(TMConstants.TMR_PARAM_REGION_SUPPORTEDREGIONS);
                if (supportedRegions.length < 1)
                {
                    throw new Exception("Reader doesn't support any regions");
                }
                else
                {
                    reader.paramSet("/reader/region/id", supportedRegions[0]);
                }
            }
//          Set tag password one time for each tag.  In real life, this occurs
//          before the tag reaches the end user.
//
//          short newpassword[] = new short[]{(short) 0x8888, (short) 0x7777};
//          TagReadData[] tags = reader.read(1000);
//          String gen2_read_tag_epc = getEPC(tags);
//          TagFilter filterToUse = new TagData(gen2_read_tag_epc);
//          Gen2.WriteData write = new Gen2.WriteData(Gen2.Bank.RESERVED, 2, newpassword);
//          reader.executeTagOp(write, filterToUse);
//
//          Gen2.ReadData rOp1 = new Gen2.ReadData(Gen2.Bank.RESERVED, 0, (byte)0);
//          short[] rdata = (short[]) reader.executeTagOp(rOp1, filterToUse);

            //Embedded Secure Read Tag Operation - Standalone operation not supported
            Gen2.Password password = new Gen2.Password(0);
            Gen2.SecureReadData secureReadDataTagOp = new Gen2.SecureReadData(Gen2.Bank.TID, 0, (byte) 0, Gen2.SecureTagType.HIGGS3, password);

            
            TagProtocol[] protocolList = (TagProtocol[]) reader.paramGet("/reader/version/supportedProtocols");
            ReadPlan setSecureReadPlan[] = new ReadPlan[protocolList.length];
            for (int i = 0; i < protocolList.length; i++)
            {
                setSecureReadPlan[i] = new SimpleReadPlan(new int[]{1, 2}, protocolList[i], null, secureReadDataTagOp, 100);
            }
            MultiReadPlan setSecureMultiReadPlan = new MultiReadPlan(setSecureReadPlan);
            reader.paramSet("/reader/read/plan", setSecureMultiReadPlan);

            // Create and add tag read authentication listener
            ReadAuthenticationListener authenticationListener = new TagReadAuthentication();
            reader.addReadAuthenticationListener(authenticationListener);
            // Create and add tag read exception listener
            ReadExceptionListener exceptionListener = new TagReadExceptionReceiver();
            reader.addReadExceptionListener(exceptionListener);
            // Create and add tag read listener
            ReadListener readListener = new PrintListener();
            reader.addReadListener(readListener);

            // search for tags in the background
            reader.startReading();

            System.out.println("Do other work here");
            Thread.sleep(600);
             System.out.println("Do other work here");
            Thread.sleep(500);
            reader.stopReading();

            //Remove listeners after stoping reading
            reader.removeReadListener(readListener);
            reader.removeReadExceptionListener(exceptionListener);
            reader.removeReadAuthenticationListener(authenticationListener);

            // Shut down reader
            reader.destroy();
        } catch (ReaderException re) {
            System.out.println("ReaderException: " + re.getMessage());
        } catch (Exception re) {
            System.out.println("Exception: " + re.getMessage());
        }
    }

    static class PrintListener implements ReadListener
    {
        public void tagRead(Reader r, TagReadData tr)
        {
            System.out.println("Background read: " + tr.toString());
        }
    }

    static class TagReadExceptionReceiver implements ReadExceptionListener
    {
        String strDateFormat = "M/d/yyyy h:m:s a";
        SimpleDateFormat sdf = new SimpleDateFormat(strDateFormat);

        public void tagReadException(com.thingmagic.Reader r, ReaderException re)
        {
            String format = sdf.format(Calendar.getInstance().getTime());
            System.out.println("Reader Exception: " + re.getMessage() + " Occured on :" + format);
            if (re.getMessage().equals("Connection Lost"))
            {
                System.exit(1);
            }
        }
    }

    static class TagReadAuthentication implements ReadAuthenticationListener
    {
        public void readTagAuthentication(TagReadData t, Reader reader) throws ReaderException
        {
            try
            {
                int accessPassword = 0;
                int tagAccessPassword1 = 0x11111111;
                int tagAccessPassword2 = 0x22222222;
                int tagAccessPassword3 = 0x33333333;
                int tagAccessPassword5 = 0x55555555;
                System.out.println("Tag epc  : " + t.epcString());
                Double tagEPC = Double.parseDouble(t.epcString());
                Double index = tagEPC % 4;
                if (index == 0) {
                    accessPassword = tagAccessPassword5;
                } else if (index == 1) {
                    accessPassword = tagAccessPassword1;
                } else if (index == 2) {
                    accessPassword = tagAccessPassword2;
                } else if (index == 3) {
                    accessPassword = tagAccessPassword3;
                }
                //Set access posword  for corresponding tag epc 
                Gen2.Password authPassword = new Gen2.Password(accessPassword);
                reader.paramSet(TMConstants.TMR_PARAM_GEN2_ACCESSPASSWORD, authPassword);
            }
            catch (ReaderException re)
            {
                re.printStackTrace();
            }
        }
    }
}

File: lib/mercuryapi-1.23.0.20/cs/Samples/Codelets/SecureReadData/SecureReadData.cs
using System;
using System.Collections.Generic;
using System.Text;
// for Thread.Sleep
using System.Threading;

// Reference the API
using ThingMagic;

namespace SecureReadData
{
    /// <summary>
    /// Sample program that supports PSAM functionality
    /// </summary>
    class SecureReadData
    {
        static void Main(string[] args)
        {
            // Program setup
            if (1 != args.Length)
            {
                Console.WriteLine(String.Join("\r\n", new string[] {
                    "Please provide reader URL, such as:",
                    "tmr:///com4",
                    "tmr://my-reader.example.com",
                }));
                Environment.Exit(1);
            }

            try
            {
                // Create Reader object, connecting to physical device.
                // Wrap reader in a "using" block to get automatic
                // reader shutdown (using IDisposable interface).
                using (Reader r = Reader.Create(args[0]))
                {
                    //Uncomment this line to add default transport listener.
                    //r.Transport += r.SimpleTransportListener;

                    r.Connect();
                    if (Reader.Region.UNSPEC == (Reader.Region)r.ParamGet("/reader/region/id"))
                    {
                        Reader.Region[] supportedRegions = (Reader.Region[])r.ParamGet("/reader/region/supportedRegions");
                        if (supportedRegions.Length < 1)
                        {
                            throw new FAULT_INVALID_REGION_Exception();
                        }
                        else
                        {
                            r.ParamSet("/reader/region/id", supportedRegions[0]);
                        }
                    }
                    // Read Plan
                    byte length;
                    string model = (string)r.ParamGet("/reader/version/model");
                    if ("M6e".Equals(model)
                        || "M6e PRC".Equals(model)
                        || "M6e Micro".Equals(model)
                        || "Mercury6".Equals(model)
                        || "Astra-EX".Equals(model))
                    {
                        // Specifying the readLength = 0 will return full TID for any tag read in case of M6e varients, M6 and Astra-EX reader.
                        length = 0;
                    }
                    else
                    {
                        length = 2;
                    }
                    
                    // Embedded Secure Read Tag Operation - Standalone operation not supported
                    Gen2.Password password = new Gen2.Password(0);
                    Gen2.SecureReadData secureReadDataTagOp = new Gen2.SecureReadData(Gen2.Bank.TID, 0, (byte)0, Gen2.SecureTagType.HIGGS3, password); ;
                    SimpleReadPlan plan = new SimpleReadPlan(null, TagProtocol.GEN2, null, secureReadDataTagOp, 1000);
                    r.ParamSet("/reader/read/plan", plan);

                    // Create and add tag listener
                    r.TagRead += delegate(Object sender, TagReadDataEventArgs e)
                    {
                        Console.WriteLine("Background read: " + e.TagReadData);
                        Console.WriteLine("Requested data: "+ByteFormat.ToHex(e.TagReadData.Data));
                    };

                    // Create and add read exception listener
                    r.ReadException += r_ReadException;

                    // Create and add read authenticate listener
                    r.ReadAuthentication += r_ReadAuthenticationListener;

                    // Search for tags in the background
                    r.StartReading();

                    Console.WriteLine("\r\n<Do other work here>\r\n");
                    Thread.Sleep(500);
                    Console.WriteLine("\r\n<Do other work here>\r\n");
                    Thread.Sleep(500);

                    r.StopReading();
                }
            }
            catch (ReaderException re)
            {
                Console.WriteLine("Error: " + re.Message);
                Console.Out.Flush();
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error: " + ex.Message);
            }
        }

        static void r_ReadAuthenticationListener(object sender, ReadAuthenticationEventArgs e)
        {
            // This example simply uses a precompiled password.
            // In real life, an algorithm or server would examine the tag ID contained in trd and return an appropriate password.
            Reader rd = e.TagReadData.Reader;
            uint accessPassword = 0;
            uint tagAccessPassword1 = 0x11223344;
            uint tagAccessPassword2 = 0x22222222;
            uint tagAccessPassword3 = 0x33333333;
            uint tagAccessPassword4 = 0x11111111;
            Console.WriteLine("Tag epc  : " + e.TagReadData.EpcString);
            int index = e.TagReadData.Epc.Length % 4;
            if (index == 0)
            {
                accessPassword = tagAccessPassword1;
            }
            else if (index == 1)
            {
                accessPassword = tagAccessPassword2;
            }
            else if (index == 2)
            {
                accessPassword = tagAccessPassword3;
            }
            else if (index == 3)
            {
                accessPassword = tagAccessPassword4;
            }
            // Set password
            rd.ParamSet("/reader/gen2/accessPassword", new Gen2.Password(accessPassword));
        }

        static void r_ReadException(object sender, ReaderExceptionEventArgs e)
        {
            Console.WriteLine("Error: " + e.ReaderException.Message);
        }
    }
}
